### DISCUSSION TITLE
SelectField JSON serialization

### DISCUSSION THREAD

**Discussion by gmullernh** [PiranhaCMS/piranha.core#1680] [Q&A]
Hello!

I'm using Newtonsoft.JSON to serialize the a Content Page as a JSON object (for headless CMS use).
When trying to serialize a SelectField with a custom Enum within a Region, it's always returning null, even with serialization attributes at the Enum.

```Csharp
public class MyPage: Page<MyPage> 
{
    public class MyRegion
    {
        [Field(Title = "My Category"), ]
        public SelectField<MyCategoryEnum> MyCategory { get; set; }
    }

    [Region(Title="Information")]
    public MyRegion Information { get; set; }
}

[JsonConverter(typeof(StringEnumConverter))]
[Serializable]
public enum MyCategoryEnum
{
    [EnumMember, Display(Description = "Value A")]
    ValueA,
    [EnumMember, Display(Description = "Value B")]
    ValueB,
}
```

JSON response:

```JSON
{
    "MyRegion": {
      "MyCategory ": null,
    }
}
```

Any ideas of what might be causing this behavior?

**Comment by tidyui** [comment]
Just to double check, does MyRegion have more fields in your application? The reason for asking is that model transformation doesn’t support object regions with a single field. 

https://piranhacms.org/docs/master/content/regions

**Reply by gmullernh** [reply]
Yes, it does contain some NumberFields, HtmlFields, CheckboxFields and DateFields, in this particular case, this region contains 15 fields.

**Reply by gmullernh** [reply]
I took a look to identify if this was an issue with asynchronicity, but it seems not. 
Checking at Visual Studio debug tool, the SelectFields are returning as `null` from the API, while the other types are returning as their respective contents like `Piranha.Extend.Fields.HtmlField`, `Piranha.Extend.Fields.StringField`, and so on.

The content is being fetch by an async method, using the API with
```Csharp
await _api.Pages.GetByIdAsync<MyPage>(pageId).ConfigureAwait(false);
```

**Reply by gmullernh** [reply]
The saga continues. Database is fine. The `CLRType` and `Value` are being registered by the manager and the issue seems to be only when the Piranha's API is fetching the enum values.
Temporarily solved by querying directly using the database context and creating a new `SelectField<MyCategoryEnum>` on the controller action.

Results from the database at the table `Piranha_PageFields`:
_GUIDs are working fine, just hidden._
```
Id: XXXXXXXXXXXX
CLRType: Piranha.Extend.Fields.SelectField`1[MyProject.Models.Pages.MyCategoryEnum]
PageId: XXXXXXXXXX
FieldId: MyCategory
RegioId: MyRegion
SortOrder: 0
Value: ValueA
```

The endpoint workaround:

```Csharp
public async Task<IActionResult> GetAsync()
{
    List<MyPage> pages = new();

    foreach (Guid pageId in _db.Pages
    .Where(p => p.Published <= DateTime.Now && p.PageTypeId == "MyPage")
    .Select(p => p.Id)
    .ToList())
    {
        MyPage page = await _api.Pages.GetByIdAsync<MyPage>(pageId).ConfigureAwait(false);
        page.MyRegion.MyCategoryEnum = new SelectField<MyCategoryEnum>() { Value = (MyCategoryEnum)Enum.Parse(typeof(MyCategoryEnum), _db.PageFields.Where(pf => pf.PageId == pageId && pf.FieldId == "MyCategory").FirstOrDefault().Value) };
        cursos.Add(page);
    }
    return Content(JsonConvert.SerializeObject(cursos));
}
```

Noticed that the API endpoint is returning the enum value as the integer value that represents the selected field, ignoring the `[JsonConverter(typeof(StringEnumConverter))]`.

```JSON
{
    "MyRegion": {
      "MyCategory ": 1,
    }
}
```

Seems to be something with how Piranha CMS recovers the enum content and how this content is being serialized.

Also, something I've noticed during tests with `SelectField<MyCategoryEnum>`. If I move the `MyCategoryEnum` to inside the MyPage class or to somewhere else, change the references so everything works within the code, when loading content in Piranha CMS Manager, those pages already created previously and that use the `SelectField<MyCategoryEnum>`, don't load anymore, throwing a 500 error from a `NullReferenceException: Object reference not set to an instance of an object` instead of just excluding the field from the manager (like other types of fields do).

Stack from `/manager/api/page/34b62185-fc41-494c-af4a-39c785f9acb0`:
```
Piranha.Services.ContentFactory.InitFieldAsync(IServiceScope scope, object field, bool managerInit)
Piranha.Services.ContentFactory.InitDynamicRegionAsync(IServiceScope scope, object region, ContentTypeRegion regionType, bool managerInit)
Piranha.Services.ContentFactory.InitDynamicAsync<T>(T model, ContentTypeBase type, bool managerInit)
Piranha.Services.PageService.OnLoadAsync(PageBase model, bool isDraft)
Piranha.Services.PageService.GetByIdsAsync<T>(Guid[] ids)
Piranha.Services.PageService.GetByIdAsync<T>(Guid id)
Piranha.Manager.Services.PageService.GetById(Guid id, bool useDraft)
Piranha.Manager.Controllers.PageApiController.Get(Guid id)
```
